llvm/lib/Target/AIE/AIEPostSelectOptimize.cpp

```diff
@@ -438,6 +440,140 @@ bool duplicateAdressingRegs(MachineBasicBlock &MBB, MachineRegisterInfo &MRI) {
+                  SmallPtrSet<const Value *, 4> &GVSet) {
+
+  bool Success = true;
+  for (const Register Reg : VSelect.VectorSrcRegs) {
+    OptionalVecOp VOp = TII->parseTargetVectorOp(*MRI.getVRegDef(Reg));
+    // We don't recognize this! Here we cannot have anything different
+    // from SELECT or BROADCAST.
+    if (!VOp)
+      return false;
+    if (VOp->Type == Operation::SELECT)
+      Success &= visitVSelect(*VOp, TII, MRI, GVSet);
+    else if (VOp->Type == Operation::BROADCAST)
+      Success &= visitVBroadcast(*VOp, TII, MRI, GVSet);
+    else
+      return false;
+  }
+
+  return Success;
+}
+
+// Recognize ADD ((ADD | SELECT), Unknown)
+bool visitVAdd(const AbstractVecOp &VAdd, const AIEBaseInstrInfo *TII,
+               MachineRegisterInfo &MRI, SmallPtrSet<const Value *, 4> &GVSet) {
+
+  bool UnknownSide = false;
+  bool KnownSide = false;
+
+  for (const Register Reg : VAdd.VectorSrcRegs) {
+    OptionalVecOp VOp = TII->parseTargetVectorOp(*MRI.getVRegDef(Reg));
+
+    // We don't recognize this, but this represents the other side of the
+    // add (offset). It can be anything, but if it leads to out-of-bounds,
+    // we don't care bacause it is UB anyway.
+    if (!VOp) {
+      UnknownSide = true;
+      continue;
+    }
+    // The pointer-side of the add.
+    if (VOp->Type == Operation::ADD)
+      KnownSide = visitVAdd(*VOp, TII, MRI, GVSet);
+    else if (VOp->Type == Operation::SELECT)
+      KnownSide = visitVSelect(*VOp, TII, MRI, GVSet);
+    else
+      return false;
+  }
+  return UnknownSide /*offset side is found*/ &&
+         KnownSide /*broadcasted side is found*/;
+}
+
+// Follow the chain up starting for ADD or directly from SELECT.
+bool traverseVecPointerChain(Register Reg, SmallPtrSet<const Value *, 4> &GVSet,
+                             const AIEBaseInstrInfo *TII,
+                             MachineRegisterInfo &MRI) {
+  const MachineInstr *MI = MRI.getVRegDef(Reg);
+  // Skip subvector copy.
+  if (MI->isCopy() && MI->getOperand(1).getReg().isVirtual())
+    MI = MRI.getVRegDef(MI->getOperand(1).getReg());
+
+  OptionalVecOp VOp = TII->parseTargetVectorOp(*MI);
+  if (!VOp)
+    return false;
+
+  if (VOp->Type == Operation::ADD)
+    return visitVAdd(*VOp, TII, MRI, GVSet);
+
+  if (VOp->Type == Operation::SELECT)
+    return visitVSelect(*VOp, TII, MRI, GVSet);
+
+  return false;
+}
+
+/// This optimization tries to propagate GlobalValues as MMO of load operations
+/// when they are missing.
+bool fixLoadMemOpInfo(MachineFunction &MF, MachineBasicBlock &MBB,
+                      MachineRegisterInfo &MRI) {
+
+  const TargetSubtargetInfo &ST = MBB.getParent()->getSubtarget();
+  const AIEBaseInstrInfo *TII =
+      static_cast<const AIEBaseInstrInfo *>(ST.getInstrInfo());
+
+  bool Changed = false;
+
+  for (MachineInstr &MI : MBB) {
+    // We want to analyze just loads without post increment.
+    if (!MI.mayLoad() || !MI.memoperands_empty() || MI.getNumOperands() != 2)
```

Comment:
This is an attempt to recognise a multi-way load in an architecture-neutral way. Why not add that to the set of abstract operations?